facade/src/main/java/org/jboss/pnc/facade/providers/BuildProviderImpl.java

```diff
@@ -302,4 +324,192 @@ public void setDependentArtifacts(int id, List<Integer> artifactIds) {
         buildRecord.setDependencies(artifacts);
         repository.save(buildRecord);
     }
+
+    /**
+     * Returns the page of builds filtered by given BuildPageInfo parameters and predicate.
+     */
+    private Page<Build> getBuildList(BuildPageInfo pageInfo, java.util.function.Predicate<BuildTask> predicate, Predicate<BuildRecord> dbPredicate) {
+        if (pageInfo.isRunning()) {
+            if (pageInfo.isLatest()) {
+                return getLatestRunningBuild(predicate);
+            } else {
+                return getRuningBuilds(pageInfo, predicate);
+            }
+        } else {
+            if (pageInfo.isLatest()) {
+                return getLatestBuild(predicate, dbPredicate);
+            } else {
+                return getBuilds(pageInfo, predicate, dbPredicate);
+            }
+        }
+    }
+
+    /**
+     * Returns the page of Latest Running build filtered by given predicate.
+     */
+    private Page<Build> getLatestRunningBuild(java.util.function.Predicate<BuildTask> predicate) {
+        List<Build> build = readLatestRunningBuld(predicate)
+                .map(Collections::singletonList)
+                .orElse(Collections.emptyList());
+        return new Page<>(0, 1, build.size(), build.size(), build);
+    }
+
+    /**
+     * Returns the page of Running builds filtered by given BuildPageInfo and predicate.
+     */
+    private Page<Build> getRuningBuilds(BuildPageInfo pageInfo, java.util.function.Predicate<BuildTask> predicate) {
+        List<Build> runningBuilds = readRunningBuilds(pageInfo, predicate);
+
+        List<Build> builds = runningBuilds.stream()
+                .skip(pageInfo.getPageIndex() * pageInfo.getPageSize())
+                .limit(pageInfo.getPageSize())
+                .collect(Collectors.toList());
+        return new Page<>(
+                pageInfo.getPageIndex(),
+                pageInfo.getPageSize(),
+                (int) Math.ceil((double) runningBuilds.size() / pageInfo.getPageSize()),
+                runningBuilds.size(),
+                builds);
+    }
+
+    /**
+     * Returns the page of Latest build (running or finished) filtered by given predicate.
+     */
+    private Page<Build> getLatestBuild(java.util.function.Predicate<BuildTask> predicate, Predicate<BuildRecord> dbPredicate) {
+        TreeSet<Build> sorted = new TreeSet<>(Comparator.comparing(Build::getSubmitTime).reversed());
+        readLatestRunningBuld(predicate).ifPresent(sorted::add);
+        readLatestFinishedBuild(dbPredicate).ifPresent(sorted::add);
+        if (sorted.size() > 1) {
+            sorted.pollLast();
```

Comment:
I need a collection containing the first element. It could be done without the if by
```
Collections.singleton(sorted.first());
```
Do you find this more readable?

Reply:
ok, that's not possible, because there may not be any present and first() would throw exception in that case.